Creating API for requests to ENS

jamiehewitt15 opened the pull request:
Fixes https://github.com/oceanprotocol/market/issues/1636

Changes proposed in this PR:

- 
- 
-

kremalicious reviewed (commented):
Looks solid in terms of functionality and response specs.

Setup needs to be cleaned a bit, too many unused dependencies, like all the babel stuff, urql, and more, overkill gitignore, unused docker build tasks, and wrong project name in package.json

kremalicious reviewed (commented):
For requests we could simplify as the `?accountId=` is kinda redundant

- `/name/0xxx`
- `/address/winstonwolfe.eth`
- `/profile/0xxxx`

kremalicious commented on tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+{
```
Are you sure all these compiler settings are needed for this Vercel project? They probably have some default one somewhere and pretty sure it will be super short

jamiehewitt15 replied:
Yeah good point, probably not all needed. I'll have a look into

jamiehewitt15 replied:
I've swapped it to use the default that vercel creates, which isn't too different

kremalicious replied:
To me this looks like the default Next.js tsconfig, and again, this is not a Next.js app. Pretty sure we don’t need more than this, and we especially should not define any custom `outDir` https://github.com/oceanprotocol/dockerhub-proxy/blob/main/tsconfig.json

jamiehewitt15 replied:
Yep, I've simplified this

kremalicious commented on pages/api/_utils.ts:
```diff
@@ -0,0 +1,71 @@
+  OperationContext,
+  fetchExchange
+} from 'urql'
+import { refocusExchange } from '@urql/exchange-refocus'
```
This is a server, there is no user refocusing the browser window, so this is useless in this context

kremalicious commented on pages/api/text.ts:
```diff
@@ -0,0 +1,62 @@
+export async function getEnsTextRecords(
+  ensName: string
+): Promise<{ key: string; value: string }[]> {
+  // 1. Check which text records are set for the domain with ENS subgraph,
```
Right, I remember. That's actually bad as the whole flow of getting text records is made dependent on this call but most likely there is no alternative yet until new ens.js is done if we want to save contract calls. Basically problem is that we need to call all text record fields to figure out if it has been set or not.

We could kick out the initial subgraph call and see how much stress this puts on our infura

kremalicious replied:
This might apply here too so it could be this would even stop working in October https://twitter.com/Mardeni01/status/1560285524920860672

jamiehewitt15 replied:
I've removed the subgraph call for now but I can put it back if it places to many requests on Infura

kremalicious commented on .env.example:
```diff
@@ -0,0 +1 @@
+#NEXT_PUBLIC_INFURA_PROJECT_ID="xxx"
```
`NEXT_PUBLIC` prefix is not needed here, right?

jamiehewitt15 replied:
Yeah, it's not needed. I'll remove it to avoid any confusion

kremalicious commented on next-env.d.ts:
```diff
@@ -0,0 +1,5 @@
+/// <reference types="next" />
```
This is not a Next.js app so is this file needed?

jamiehewitt15 replied:
This file is automatically generated when you run `vercel dev`. It's because we're using the [next.js API routes](https://nextjs.org/docs/api-routes/introduction)

kremalicious replied:
Again, this is a Vercel serverless function, and NOT a Next.js app and we are not putting things in `/pages/api`. This is the documentation we are using https://vercel.com/docs/concepts/functions/serverless-functions

jamiehewitt15 replied:
Everything related to next.js has been removed now

kremalicious commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "format": "prettier --ignore-path .gitignore './**/*.{css,yml,js,ts,tsx,json}' --write",
+    "test:format": "npm run format && npm run lint",
+    "type-check": "tsc --noEmit",
+    "test:integration": "mocha --config=test/.mocharc.json --node-env=test --exit 'test/**/*.test.ts'",
```
I do not see any test folder

jamiehewitt15 replied:
I've added tests now

kremalicious commented on package.json:
```diff
@@ -0,0 +1,40 @@
+    "prettier": "^2.7.1",
+    "pretty-quick": "^3.1.3",
+    "supertest": "^6.2.3",
+    "ts-node": "^10.8.1",
```
That's unused, same for `supertest`, `mocha`, `chai`

kremalicious commented on pages/api/text.ts:
```diff
@@ -0,0 +1,62 @@
+      requestPolicy: 'cache-and-network'
+    }
+  )
+  if (!result?.data?.domains[0]?.resolver) throw 'No ENS text records found'
```
Note for later when I have figured out Vercel deployment of this project: we need to make sure if nothing is found or set, we do not error out. Nothing found should be `200` repsonse too, as request and response actually works. So rather we probably should have a specific response if nothing is found

jamiehewitt15 replied:
Yeah, I've adjusted this so that 200 is always sent whether or not something is found. This is also part of the tests

kremalicious commented on pages/api/address.ts:
```diff
@@ -0,0 +1,19 @@
+    const address = await ens.name(ensName).getAddress()
+    if (!address) throw `No address found for "${ensName}"`
+
+    response.setHeader('Cache-Control', 's-maxage=86400')
```
Vercel recommends to also add `max-age=0` as this deals with cache invalidation for redeployments as they write https://vercel.com/docs/concepts/functions/serverless-functions/edge-caching

kremalicious commented on README.md:
````diff
@@ -1 +1,96 @@
+### Get Ens Name
+
+```
+GET http://localhost:3000/api/name?accountId=0x903322c7e45a60d7c8c3ea236c5bea9af86310c7
````
Let's use the jellymcjellyfish.eth account as example account here too

kremalicious commented on pages/api/_utils.ts:
```diff
@@ -0,0 +1,71 @@
+
+let ens: any
+
+function getOceanConfig(network: string | number): Config {
```
This is overengineered. ENS only works on ETH mainnet. So we know we only have to use the mainnet Infura URL. Whioch we simply define as a `const` in the below `getDummyWeb3()`, which itself should rather be named `getWeb3()`

kremalicious replied:
This is still over engineered, you are loading all of ocean.js to get a widely known Infura URL. Remove the need to have ocean.js as dependency

kremalicious commented on .eslintrc:
```diff
@@ -0,0 +1,33 @@
+{
+  "parser": "@babel/eslint-parser",
```
This is not among dependencies, and this is a TypeScript project without a single js file so this is not needed to be configured here with ESLint. We only want the typescript eslint parser for everything

kremalicious commented on pages/api/_utils.ts:
```diff
@@ -0,0 +1,71 @@
+import { ConfigHelper, Config, LoggerInstance } from '@oceanprotocol/lib'
+import Web3 from 'web3'
```
As I have written in the original PR on market, if we can, we should rather use ethers.js. Then we do not need to use this old ens.js library which is due to be replaced with a new major version. In market we can't just use another web3 library, but here we have full flexibility about which libraries to use.

Ethers.js has multiple methods included which I tried to replicate with web3.js. So look into ethers.js and see what code we can ditch if we use that https://docs.ethers.io/v5/api/providers/provider/#Provider--ens-methods

jamiehewitt15 replied:
I've replaced it with ethers which seems to work nicely

kremalicious commented on pages/api/address.ts:
```diff
@@ -0,0 +1,19 @@
+import { NextApiRequest, NextApiResponse } from 'next'
```
As this is not a Next.js project, we should not use those packages. Please use the appropriate packages you can find e.g. here https://vercel.com/docs/concepts/functions/serverless-functions/supported-languages#using-typescript

kremalicious commented on .eslintrc:
```diff
@@ -0,0 +1,32 @@
+      "version": "detect"
+    }
+  },
+  "overrides": [
```
there should be no override if we only use 1 parser. So remove this override and make ESLint only use `@typescript-eslint/parser`

kremalicious commented:
And still have not managed to actually make the API work. How is this supposed to work remotely?

This gives me nothing:

```
curl https://ens-proxy-a02wg6da2-oceanprotocol.vercel.app/api/name?accountId=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
```

Pretty sure this PR putting things under `/pages/api` is the problem.

jamiehewitt15 commented:
The automatic deployment is working now so you can test with any of the following:

- Get the ENS name
```
curl https://ens-proxy-7bvvswjla-oceanprotocol.vercel.app/api/name?accountId=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
```

 - Get the ENS address
```
curl https://ens-proxy-7bvvswjla-oceanprotocol.vercel.app/api/address?name=jellymcjellyfish.eth
```

 - Get the ENS text records
```
curl https://ens-proxy-7bvvswjla-oceanprotocol.vercel.app/api/text?name=jellymcjellyfish.eth
```

 - Get the profile
```
curl https://ens-proxy-7bvvswjla-oceanprotocol.vercel.app/api/profile?address=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
```

EnzoVezzaro commented:
All endpoints are reachable. I've created a [postman snapshot](https://www.getpostman.com/collections/e3de32b11efc261487fb) in case you guys need. 

Sorry if I'm saying something already discussed, but I can comment on the responses here:

I'd expect the response to have some sort of JSON specification. Right now we have these responses:

- Get the ENS name **return a String**
- Get the ENS address **return a String**
- Get the ENS text records **return a Object**
- Get the profile **return a Object**

I would personally use something like Google JSON style guide to the responses, for example: 

Success response:
`
{
  "data": {
    ...
  }
}

`

Error response:
`
{
  "error": {
    "code": 404,
    "message": "..."
  }
}
`

or some variation of that (maybe adding also the status code) in order to always have the same pattern in the response. 

This way we can catch and throw the error messages in the response. Right now, 3 out of 4 endpoint are returning error 500 on wring input:

- Get the ENS name (example: `0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0d `)
<img width="1008" alt="Screenshot 2022-08-23 at 10 47 13" src="https://user-images.githubusercontent.com/13741335/186189122-52029127-a12b-4703-bc49-cefda075d5e7.png">

- Get the ENS address (example: `jellymcjellyfish.ethr `)
<img width="558" alt="Screenshot 2022-08-23 at 10 48 02" src="https://user-images.githubusercontent.com/13741335/186189295-71dfd290-4d65-4e91-b4d0-03f66868d17c.png">

- Get the ENS text records (example: `jellymcjellyfish.ethf `)
<img width="609" alt="Screenshot 2022-08-23 at 10 47 34" src="https://user-images.githubusercontent.com/13741335/186189186-7664cbab-c4dc-4942-a9de-53724c7ce696.png">

- Get the profile (example: `0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0f `)
<img width="1021" alt="Screenshot 2022-08-23 at 10 47 48" src="https://user-images.githubusercontent.com/13741335/186189231-6193a112-8783-4779-b32b-df6438b8de5c.png">

jamiehewitt15 commented:
Thanks for the feedback @EnzoVezzaro, I've updated the responses so that every endpoint returns an object:

### Get Ens Name

```
GET https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/name?accountId=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
```

Example response:

```
{
    "name": "jellymcjellyfish.eth"
}
```

### Get Ens Address

```
GET https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/address?name=jellymcjellyfish.eth
```

Example response:

```
{
    "address": "0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0"
}
```

### Get Ens Text Records

```
GET https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/text?name=jellymcjellyfish.eth
```

Example response:

```
{
    "records": [
        {
            "key": "url",
            "value": "https://oceanprotocol.com"
        },
        {
            "key": "avatar",
            "value": "https://raw.githubusercontent.com/oceanprotocol/art/main/logo/favicon-white.png"
        },
        {
            "key": "com.twitter",
            "value": "oceanprotocol"
        },
        {
            "key": "com.github",
            "value": "oceanprotocol"
        }
    ]
}
```

### Get ENS Profile

```
GET https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/profile?address=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
```

Example response:

```
{
    "profile": {
        "name": "jellymcjellyfish.eth",
        "avatar": "https://metadata.ens.domains/mainnet/avatar/jellymcjellyfish.eth",
        "links": [
            {
                "key": "url",
                "value": "https://oceanprotocol.com"
            },
            {
                "key": "com.twitter",
                "value": "oceanprotocol"
            },
            {
                "key": "com.github",
                "value": "oceanprotocol"
            }
        ]
    }
}
```

I've also updated the error handling so that it doesn't send 500 errors now when the address isn't found. These responses are now in the following format:

### Invalid address:

```
https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/profile?address=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6Cdd84e0
```

```
{
    "error": {
        "reason": "invalid address",
        "code": "INVALID_ARGUMENT",
        "argument": "address",
        "value": "0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6Cdd84e0"
    }
}
```

### No results for ens name

```
https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/address?name=jellymcjellxxxyfish.eth
```

```
{
    "address": null
}
```

EnzoVezzaro commented:
not sure if it's worthy, but I was thinking of something more like: 

on "Get Ens Address" instead of
`{
      "address": "0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0"
}`

we can do something like:
`{
    "status": 200,
     "data": {
             "address": "0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0"
      }
}`

on "No results for ens name" instead of
`{
    "address": null
}`

we can do something like:
`{
    "status": 200,
     "data": null
}`

For errors we can do: 

Example: 
https://ens-proxy-mmro5x4tf-oceanprotocol.vercel.app/api/name?wrongParam=0x99840Df5Cb42faBE0Feb8811Aaa4BC99cA6C84e0
`{
    "status": 404,
    "error": {
        "reason": "invalid address",   /// THIS IS WRONG, IT SHOULD BE INVALID ARGUMENT OR PARAMETER
        "code": "INVALID_ARGUMENT",
        "argument": "address",
        "value": "undefined"
    }
}`

and so on. This way users can catch the status and show the error message. 

It might be overkill, but that's how I would like it.

jamiehewitt15 commented:
> It might be overkill, but that's how I would like it.

Not sure what we gain by sending the status in the body as it is already in the header? So we would be sending it twice. 

You can access the status in the front end with `response.status`. I've added the status as part of the tests [here](https://github.com/oceanprotocol/ens-proxy/pull/1/commits/71080e8827fbd371fc0b9d8045207d247859a666) to check that it is still 200 in both situations where the ENS record is found and not found

EnzoVezzaro commented:
> You can access the status in the front end with `response.status`. I've added the status as part of the tests [here](https://github.com/oceanprotocol/ens-proxy/pull/1/commits/71080e8827fbd371fc0b9d8045207d247859a666) to check that it is still 200 in both situations where the ENS record is found and not found

Yes, I guess it works. I'm just more used to have the status on the response, but totally understand that [we don't handle many HTTPS status](https://stackoverflow.com/questions/69945370/status-code-inside-the-body-of-a-rest-api-response), so it's not really worthy.

EnzoVezzaro reviewed (approved):
LGTM 👍🏽

Maybe we could add something in index (like a welcome message) to avoid having a 404.

<img width="2546" alt="Screenshot 2022-08-26 at 08 24 19" src="https://user-images.githubusercontent.com/13741335/186902951-a3150b4b-507c-417d-85e9-09ab162f594d.png">
